hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java

```diff
@@ -0,0 +1,96 @@
+package com.hedera.services.store.tokens.unique;
+
+/*
+ * ‌
+ * Hedera Services Node
+ * ​
+ * Copyright (C) 2018 - 2021 Hedera Hashgraph, LLC
+ * ​
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.hedera.services.context.properties.GlobalDynamicProperties;
+import com.hedera.services.ledger.TransactionalLedger;
+import com.hedera.services.ledger.ids.EntityIdSource;
+import com.hedera.services.ledger.properties.TokenRelProperty;
+import com.hedera.services.state.merkle.MerkleEntityId;
+import com.hedera.services.state.merkle.MerkleToken;
+import com.hedera.services.state.merkle.MerkleTokenRelStatus;
+import com.hedera.services.state.merkle.MerkleUniqueToken;
+import com.hedera.services.state.merkle.MerkleUniqueTokenId;
+import com.hedera.services.state.submerkle.EntityId;
+import com.hedera.services.state.submerkle.RichInstant;
+import com.hedera.services.store.tokens.BaseTokenStore;
+import com.hedera.services.txns.validation.OptionValidator;
+import com.hedera.services.utils.invertible_fchashmap.FCInvertibleHashMap;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenID;
+import com.swirlds.fcmap.FCMap;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.function.Supplier;
+
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_HAS_NO_SUPPLY_KEY;
+
+/**
+ * Provides functionality to work with Unique tokens.
+ *
+ * @author Yoan Sredkov
+ */
+public class UniqueTokenStore extends BaseTokenStore implements UniqueStore {
+
+	private final Supplier<FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier>> uniqueTokensSupply;
+
+	public UniqueTokenStore(final EntityIdSource ids,
+							final OptionValidator validator,
+							final GlobalDynamicProperties properties,
+							final Supplier<FCMap<MerkleEntityId, MerkleToken>> tokens,
+							final Supplier<FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier>> uniqueTokensSupply,
+							final TransactionalLedger<Pair<AccountID, TokenID>, TokenRelProperty, MerkleTokenRelStatus> tokenRelsLedger
+	) {
+		super(ids, validator, properties, tokens, tokenRelsLedger);
+		this.uniqueTokensSupply = uniqueTokensSupply;
```

Comment:
nit: lets rename this prop to `uniqueTokens` or `uniqueTokensSupplier`. Supply is used for other things in the context of tokens and it might be misleading